fhir-server/src/main/java/au/csiro/pathling/QueryHelpers.java

```diff
@@ -141,81 +119,137 @@ public static IdAndValue updateGroupingColumns(@Nonnull final ParserContext cont
+
+    final Column joinCondition = leftId.equalTo(rightId);
     final Dataset<Row> target = selectJoinTarget(right.getDataset(), left.getDataset());
     return left.getDataset().join(target, joinCondition, joinType.getSparkName());
   }
 
   /**
-   * @param left A {@link Dataset}
-   * @param leftId The ID {@link Column} in the left Dataset
-   * @param right A {@link FhirPath} expression
-   * @param joinType A {@link JoinType}
+   * Joins a {@link Dataset} to a {@link FhirPath}, using resource identity columns.
+   * <p>
+   * This should not be used in contexts where there may be grouping columns.
+   *
+   * @param left a {@link Dataset}
+   * @param leftId the ID {@link Column} in the left Dataset
+   * @param right a {@link FhirPath} expression
+   * @param joinType a {@link JoinType}
    * @return A new {@link Dataset}
    */
   @Nonnull
   public static Dataset<Row> joinOnId(@Nonnull final Dataset<Row> left,
       @Nonnull final Column leftId, @Nonnull final FhirPath right,
       @Nonnull final JoinType joinType) {
-    // Don't do unnecessary joins between identical datasets.
-    if (left.equals(right.getDataset())) {
-      return left;
-    }
+    return joinOnId(left, leftId, right, Optional.empty(), joinType);
+  }
+
+  /**
+   * Joins a {@link Dataset} to a {@link FhirPath}, using resource identity columns.
+   * <p>
+   * This should not be used in contexts where there may be grouping columns.
+   *
+   * @param left a {@link Dataset}
+   * @param leftId the ID {@link Column} in the left Dataset
+   * @param right a {@link FhirPath} expression
+   * @param additionalCondition an additional condition that will be combined with the ID equality
+   * using AND
+   * @param joinType a {@link JoinType}
+   * @return A new {@link Dataset}
+   */
+  @Nonnull
+  public static Dataset<Row> joinOnId(@Nonnull final Dataset<Row> left,
+      @Nonnull final Column leftId, @Nonnull final FhirPath right,
+      @Nonnull final Optional<Column> additionalCondition, @Nonnull final JoinType joinType) {
     check(right.getIdColumn().isPresent());
-    final Column joinCondition = leftId.equalTo(right.getIdColumn().get());
+    Column joinCondition = leftId.equalTo(right.getIdColumn().get());
+    if (additionalCondition.isPresent()) {
+      joinCondition = joinCondition.and(additionalCondition.get());
+    }
     final Dataset<Row> target = selectJoinTarget(right.getDataset(), left);
     return left.join(target, joinCondition, joinType.getSparkName());
   }
 
   /**
-   * @param left A {@link Dataset}
-   * @param leftId The ID {@link Column} in the left Dataset
-   * @param right Another Dataset
-   * @param rightId The ID column in the right Dataset
-   * @param joinType A {@link JoinType}
-   * @return A new {@link Dataset}
+   * Joins a {@link Dataset} to another Dataset, using resource identity columns.
+   * <p>
+   * This should not be used in contexts where there may be grouping columns.
```

Comment:
There does not seem to be check for this (no grouping columns) anywhere.

Reply:
We do check for the existence of the resource identity column, but we can't check for the grouping columns here as we don't have the context (which is where the grouping columns live).

I do agree that this stuff is a bit messy, but I haven't thought of a better way to do it as yet.